src/main/java/io/gravitee/policy/groovy/GroovyPolicy.java

```diff
@@ -15,250 +15,137 @@
+import io.gravitee.gateway.reactive.api.message.Message;
+import io.gravitee.gateway.reactive.api.policy.Policy;
+import io.gravitee.policy.groovy.PolicyResult.State;
 import io.gravitee.policy.groovy.configuration.GroovyPolicyConfiguration;
-import io.gravitee.policy.groovy.model.ContentAwareRequest;
-import io.gravitee.policy.groovy.model.ContentAwareResponse;
-import io.gravitee.policy.groovy.sandbox.SecuredGroovyShell;
-import io.gravitee.policy.groovy.utils.AttributesBasedExecutionContext;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
+import io.gravitee.policy.groovy.model.GroovyBindings;
+import io.gravitee.policy.v3.groovy.GroovyPolicyV3;
+import io.reactivex.rxjava3.core.Completable;
+import io.reactivex.rxjava3.core.Flowable;
+import io.reactivex.rxjava3.core.Maybe;
+import lombok.extern.slf4j.Slf4j;
 
 /**
- * @author David BRASSELY (david.brassely at graviteesource.com)
+ * @author Antoine CORDIER (antoine.cordier at graviteesource.com)
  * @author GraviteeSource Team
  */
-public class GroovyPolicy {
-
-    private static final Logger logger = LoggerFactory.getLogger(GroovyPolicy.class);
+@Slf4j
+public class GroovyPolicy extends GroovyPolicyV3 implements Policy {
 
-    private final GroovyPolicyConfiguration groovyPolicyConfiguration;
+    private static final ExecutionFailure DEFAULT_EXECUTION_FAILURE = new ExecutionFailure(INTERNAL_SERVER_ERROR_500)
+        .key("GROOVY_EXECUTION_FAILURE")
+        .message("Groovy execution failure");
 
-    private static final String REQUEST_VARIABLE_NAME = "request";
-    private static final String RESPONSE_VARIABLE_NAME = "response";
-    private static final String CONTEXT_VARIABLE_NAME = "context";
-    private static final String RESULT_VARIABLE_NAME = "result";
+    private final Flowable<String> scriptFlowable;
 
-    private static final SecuredGroovyShell GROOVY_SHELL = new SecuredGroovyShell();
+    public GroovyPolicy(GroovyPolicyConfiguration configuration) {
+        super(configuration);
+        scriptFlowable = Flowable.fromIterable(configuration.getScripts());
+    }
 
-    public GroovyPolicy(GroovyPolicyConfiguration groovyPolicyConfiguration) {
-        this.groovyPolicyConfiguration = groovyPolicyConfiguration;
+    @Override
+    public String id() {
+        return "policy-groovy";
     }
 
-    @OnRequest
-    public void onRequest(Request request, Response response, ExecutionContext executionContext, PolicyChain policyChain) {
-        executeScript(request, response, executionContext, policyChain, groovyPolicyConfiguration.getOnRequestScript());
+    @Override
+    public Completable onRequest(HttpExecutionContext ctx) {
+        return ctx
+            .request()
+            .onBody(bodyBuffer ->
+                bodyBuffer
+                    .defaultIfEmpty(Buffer.buffer())
+                    .flatMapMaybe(buffer -> onHttp(ctx, buffer, ctx.request().headers(), GroovyBindings.bindRequest(ctx, buffer)))
+            );
```

Comment:
This always loads the request body in the memory even if the Groovy script does not access it. There is probably something to do to avoid this.

Reply:
✅